msgpack/decode.ts

```diff
@@ -0,0 +1,244 @@
+  pointer: { consumed: number },
+): EncodeType {
+  const type = dataView.getUint8(pointer.consumed);
+  pointer.consumed++;
+
+  if (type <= 0x7f) { // positive fixint - really small positive number
+    return type;
+  }
+
+  if ((type & FIXMAP_MASK) === FIXMAP_BITS) { // fixmap - small map
+    const size = type & ~FIXMAP_MASK;
+    return decodeMap(uint8, dataView, size, pointer);
+  }
+
+  if ((type & FIXARRAY_MASK) === FIXARRAY_BITS) { // fixarray - small array
+    const size = type & ~FIXARRAY_MASK;
+    return decodeArray(uint8, dataView, size, pointer);
+  }
+
+  if ((type & FIXSTR_MASK) === FIXSTR_BITS) { // fixstr - small string
+    const size = type & ~FIXSTR_MASK;
+    return decodeString(uint8, size, pointer);
+  }
+
+  if (type >= 0xe0) { // negative fixint - really small negative number
+    return type - 256;
+  }
+
+  switch (type) {
+    case 0xc0: // nil
+      return null;
+    case 0xc1: // (never used)
+      throw new Error(
+        "Messagepack decode encountered a type that is never used",
+      );
+    case 0xc2: // false
+      return false;
+    case 0xc3: // true
+      return true;
+    case 0xc4: { // bin 8 - small Uint8Array
+      const length = dataView.getUint8(pointer.consumed);
+      pointer.consumed++;
+      const u8 = uint8.subarray(pointer.consumed, pointer.consumed + length);
+      pointer.consumed += length;
+      return u8;
+    }
+    case 0xc5: { // bin 16 - medium Uint8Array
+      const length = dataView.getUint16(pointer.consumed);
+      pointer.consumed += 2;
+      const u8 = uint8.subarray(pointer.consumed, pointer.consumed + length);
+      pointer.consumed += length;
+      return u8;
+    }
+    case 0xc6: { // bin 32 - large Uint8Array
+      const length = dataView.getUint32(pointer.consumed);
+      pointer.consumed += 4;
+      const u8 = uint8.subarray(pointer.consumed, pointer.consumed + length);
+      pointer.consumed += length;
+      return u8;
+    }
+    case 0xc7: // ext 8 - small extension type
+    case 0xc8: // ext 16 - medium extension type
+    case 0xc9: // ext 32 - large extension type
+      throw new Error("ext not implemented yet");
+    case 0xca: { // float 32
+      const value = dataView.getFloat32(pointer.consumed);
+      pointer.consumed += 4;
+      return value;
+    }
+    case 0xcb: { // float 64
+      const value = dataView.getFloat64(pointer.consumed);
+      pointer.consumed += 8;
+      return value;
+    }
+    case 0xcc: { // uint 8
+      const value = dataView.getUint8(pointer.consumed);
+      pointer.consumed += 1;
+      return value;
+    }
+    case 0xcd: { // uint 16
+      const value = dataView.getUint16(pointer.consumed);
+      pointer.consumed += 2;
+      return value;
+    }
+    case 0xce: { // uint 32
+      const value = dataView.getUint32(pointer.consumed);
+      pointer.consumed += 4;
+      return value;
+    }
+    case 0xcf: { // uint 64
+      const value = dataView.getBigUint64(pointer.consumed);
```

Comment:
Does this mean `integer` type in msgpack could become both `number` and `bigint` in JS? Is this the same as the other decoder such as `msgpackr`?